src/core/backbone_router/multicast_listeners_table.cpp

```diff
@@ -0,0 +1,310 @@
+    mNumValidListeners++;
+
+    FixHeap(mNumValidListeners - 1);
+
+    if (mCallback != nullptr)
+    {
+        mCallback(mCallbackContext, OT_BACKBONE_ROUTER_MULTICAST_LISTENER_ADDED, &aAddress);
+    }
+
+exit:
+    LogMulticastListenersTable("Add", aAddress, aExpireTime, error);
+    CheckInvariants();
+    return error;
+}
+
+void MulticastListenersTable::Remove(const Ip6::Address &aAddress)
+{
+    otError error = OT_ERROR_NOT_FOUND;
+
+    for (uint16_t i = 0; i < mNumValidListeners; i++)
+    {
+        Listener &listener = mListeners[i];
+
+        if (listener.GetAddress() == aAddress)
+        {
+            mNumValidListeners--;
+
+            if (i != mNumValidListeners)
+            {
+                listener = mListeners[mNumValidListeners];
+                FixHeap(i);
+            }
+
+            if (mCallback != nullptr)
+            {
+                mCallback(mCallbackContext, OT_BACKBONE_ROUTER_MULTICAST_LISTENER_REMOVED, &aAddress);
+            }
+
+            ExitNow(error = OT_ERROR_NONE);
+        }
+    }
+
+exit:
+    LogMulticastListenersTable("Remove", aAddress, static_cast<TimeMilli>(0), error);
+    CheckInvariants();
+}
+
+void MulticastListenersTable::Expire(void)
+{
+    TimeMilli    now = TimerMilli::GetNow();
+    Ip6::Address address;
+
+    while (mNumValidListeners > 0 && now >= mListeners[0].GetExpireTime())
+    {
+        LogMulticastListenersTable("Expire", mListeners[0].GetAddress(), mListeners[0].GetExpireTime(), OT_ERROR_NONE);
+        address = mListeners[0].GetAddress();
+
+        mNumValidListeners--;
+
+        if (mNumValidListeners > 0)
+        {
+            mListeners[0] = mListeners[mNumValidListeners];
+            FixHeap(0);
+        }
+
+        if (mCallback != nullptr)
+        {
+            mCallback(mCallbackContext, OT_BACKBONE_ROUTER_MULTICAST_LISTENER_REMOVED, &address);
+        }
+    }
+
+    CheckInvariants();
+}
+
+void MulticastListenersTable::LogMulticastListenersTable(const char *        aAction,
+                                                         const Ip6::Address &aAddress,
+                                                         TimeMilli           aExpireTime,
+                                                         otError             aError)
+{
+    OT_UNUSED_VARIABLE(aAction);
+    OT_UNUSED_VARIABLE(aAddress);
+    OT_UNUSED_VARIABLE(aExpireTime);
+    OT_UNUSED_VARIABLE(aError);
+
+    otLogDebgBbr("MulticastListenersTable: %s %s expire %lu: %s", aAction, aAddress.ToString().AsCString(), aExpireTime,
+                 otThreadErrorToString(aError));
+}
+
+void MulticastListenersTable::FixHeap(uint16_t aIndex)
+{
+    if (!SiftHeapElemDown(aIndex))
+    {
+        SiftHeapElemUp(aIndex);
+    }
+}
+
+void MulticastListenersTable::CheckInvariants(void) const
+{
+#if OPENTHREAD_EXAMPLES_SIMULATION
+    OT_ASSERT(std::is_heap(&mListeners[0], &mListeners[mNumValidListeners], Listener::IsAfter));
```

Comment:
##### For `CheckInvariants()`:
This code take advantage of the simulation platform to run more consistency  checks (with higher overhead).
I believe it is a good practice:
- It allows us to catch bugs more quickly, and gives us much more information when bugs are found. Unittests do cover the basic usage but might unable to cover some corner cases, or it might not be updated properly in future code refactoring. 
- It does not introduce any overhead on any platform except simulation, on which platform we don't care about performance much.

##### For `std::is_heap`
I used `std::is_heap` to avoid implementing it manually. We upgraded to C++11 standard recently so that we can take advantage of modern C++ features. Maybe we can have a further discussion on which std libraries can we use and which we can't. 
**For now, I will remove usage of `std:is_heap` and write our own version.**